query/src/main/java/com/connexta/replication/query/QueryService.java

```diff
@@ -0,0 +1,219 @@
+      return;
+    }
+
+    Set<Filter> filters = queryServiceTools.activeFiltersFor(site).collect(Collectors.toSet());
+    for (Filter filter : filters) {
+      FilterIndex filterIndex = queryServiceTools.getOrCreateFilterIndex(filter);
+
+      // holds the modified time of each task put in the queue for this filter
+      // this will be used to grab the latest modified time after
+      Set<Instant> modifiedTimes = new HashSet<>();
+
+      try {
+        getNewMetadataForFilter(filter, filterIndex)
+            .map(metadatum -> metadataToTask(metadatum, filter.getPriority()))
+            .forEach(task -> queueTaskAndPutLastModifiedInSet(task, modifiedTimes));
+      } catch (AdapterException e) {
+        LOGGER.debug(
+            "Couldn't query site {}. Stopping queries for this site until next polling attempt.",
+            site.getName(),
+            e);
+        return;
+      } catch (QueueException qe) {
+        LOGGER.debug(
+            "Failed to put a task in queue for site {}. Stopping queuing of tasks for this site. Will resume at next polling attempt.",
+            site.getName(),
+            qe);
+        return;
+      }
+      saveLastModifiedInFilterIndex(filterIndex, modifiedTimes);
+    }
+  }
+
+  private Stream<Metadata> getNewMetadataForFilter(Filter filter, FilterIndex filterIndex)
+      throws AdapterException {
+    // get last modified date
+    final Date indexDate = filterIndex.getModifiedSince().map(Date::from).orElse(null);
+    if (indexDate == null) {
+      LOGGER.trace("No index for filter {} found.", filter.getName());
+    }
+
+    // not implementing failure retries here
+    QueryRequest queryRequest =
+        new QueryRequestImpl(
+            filter.getFilter(),
+            Collections.emptyList(), // ignoring excluded nodes (origins tag)
+            Collections.emptyList(), // ignoring failed items
+            indexDate);
+
+    Iterable<Metadata> changeSet = adapter.query(queryRequest).getMetadata();
+    return StreamSupport.stream(changeSet.spliterator(), false);
+  }
+
+  private TaskInfo metadataToTask(Metadata metadata, byte filterPriority) {
+    TaskInfoImpl task = new TaskInfoImpl();
+    task.setPriority(filterPriority);
+    task.setOperation(determineOperationType());
+
+    if (metadata.getResourceSize() > 0) { // determine if there is a resource
+      final ResourceInfoImpl resourceInfo = new ResourceInfoImpl(metadata);
+      task.setResourceInfo(resourceInfo);
+    }
+
+    final MetadataInfo metadataInfo = new DdfMetadataInfoImpl("Metacard XML", metadata);
+    task.setMetadatas(Stream.of(metadataInfo));
+
+    task.setLastModified(metadata.getMetadataModified().toInstant());
+
+    return task;
+  }
+
+  private OperationType determineOperationType() {
+    // TODO: determine the operation type based on the site type and kind
+    return OperationType.HARVEST;
+  }
+
+  private void queueTaskAndPutLastModifiedInSet(TaskInfo task, Set<Instant> modifiedTimes) {
```

Comment:
✏️ Consider to simply return the task created such that you can do a `map()` instead of a `forEach()` and leave it up to the loop to update the latest modified time as opposed to scattering this logic across multiple methods.

Reply:
Not sure what you mean here. The task is created in another method and how would the loop update the latest modified time?

Reply:
I think we talked about that yesterday. What I proposed cannot be done here since we want to keep track of the latest modified timestamp as we iterate the stream otherwise when the stream is interrupted, we will loose that info.